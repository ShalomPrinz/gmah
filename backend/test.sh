#!/bin/bash

if [[ $1 == "--coverage" ]]; then
    coverage run -m unittest discover tests
    coverage html
else
    python -m unittest discover tests
fi
