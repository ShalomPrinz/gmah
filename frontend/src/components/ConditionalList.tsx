import { cloneElement } from "react";

interface ConditionalListProps {
  /** Warning: Use only when list isn't dynamic */
  indexAsKey?: boolean;
  /** Use when list is of primitive type, and every item is unique */
  itemAsKey?: boolean;
  itemCallback: (item: any) => JSX.Element;
  keyProp?: string;
  list: any[];
}

const ConditionalList = ({
  indexAsKey,
  itemAsKey,
  itemCallback,
  keyProp = "id",
  list,
}: ConditionalListProps) => (
  <>
    {list.map((item, index) => {
      if (!indexAsKey && !itemAsKey && !Object.hasOwn(item, keyProp))
        console.error("No unique id found for item", item);

      const key = itemAsKey ? item : item[keyProp] ?? index;
      return cloneElement(itemCallback(item), { key });
    })}
  </>
);

export default ConditionalList;
