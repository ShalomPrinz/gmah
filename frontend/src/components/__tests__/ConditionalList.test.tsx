import { describe, expect, it } from "vitest";

import { render } from "../../../test";

import ConditionalList from "../ConditionalList";

describe("ConditionalList", () => {
  describe("snapshot", () => {
    it("should render empty list", () => {
      const { asFragment } = render(
        <ConditionalList itemCallback={() => <></>} list={[]} />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render one item list", () => {
      const { asFragment } = render(
        <ConditionalList
          itemCallback={(item: string) => <div>{item}</div>}
          list={["item"]}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render three items list", () => {
      const { asFragment } = render(
        <ConditionalList
          itemCallback={(item: string) => <h2>{item}</h2>}
          list={["first", "second", "third"]}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should retain items props", () => {
      const { asFragment } = render(
        <ConditionalList
          itemCallback={(item: JSX.Element) => <>{item}</>}
          list={[
            <h1 className="first">first</h1>,
            <h1 className="second">second</h1>,
            <h1 className="third">third</h1>,
          ]}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
