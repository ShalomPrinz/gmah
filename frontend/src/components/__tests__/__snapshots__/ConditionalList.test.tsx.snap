// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConditionalList > snapshot > should render empty list 1`] = `<DocumentFragment />`;

exports[`ConditionalList > snapshot > should render one item list 1`] = `
<DocumentFragment>
  <div>
    item
  </div>
</DocumentFragment>
`;

exports[`ConditionalList > snapshot > should render three items list 1`] = `
<DocumentFragment>
  <h2>
    first
  </h2>
  <h2>
    second
  </h2>
  <h2>
    third
  </h2>
</DocumentFragment>
`;

exports[`ConditionalList > snapshot > should retain items props 1`] = `
<DocumentFragment>
  <h1
    class="first"
  >
    first
  </h1>
  <h1
    class="second"
  >
    second
  </h1>
  <h1
    class="third"
  >
    third
  </h1>
</DocumentFragment>
`;
